### ISSUE TITLE
better output for avocado multiplexer

### ISSUE THREAD

**Issue by jscotka** [avocado-framework/avocado#1811]
Hi
I'm using multiplexer now, and I've found that it is little bit not confortable to use it for my usecase.
My usecase is to run all tests for one test variant and then run another one.
Now it is mixed together, like test:1 , test:2

second thing is, that it would be nice to mark results like: test:prettyname, test:anotherprettyname, than test:1
```
==> default: JOB ID     : a3dc38146a3c04fc16552810af41a16475f42ee3
==> default: JOB LOG    : /root/avocado/job-results/job-2017-02-22T13.30-a3dc381/job.log
==> default:  (01/14) ./generated.py:GeneratedTestsConfig.test_processrunnig;1:  
==> default: PASS (34.05 s)
==> default:  (02/14) ./generated.py:GeneratedTestsConfig.test_processrunnig;2:  
==> default: PASS (24.25 s)
==> default:  (03/14) ./generated.py:GeneratedTestsConfig.test_selfcheck;1:  
==> default: ERROR (14.15 s)
==> default:  (04/14) ./generated.py:GeneratedTestsConfig.test_selfcheck;2:  
==> default: ERROR (2.98 s)
==> default:  (05/14) ./sanity1.py:SanityCheck1.test1;1:  
==> default: PASS (13.79 s)
==> default:  (06/14) ./sanity1.py:SanityCheck1.test1;2:  
==> default: PASS (2.86 s)
==> default:  (07/14) ./sanity1.py:SanityCheck1.test2;1:  
==> default: PASS (13.76 s)
==> default:  (08/14) ./sanity1.py:SanityCheck1.test2;2:  
==> default: PASS (2.87 s)
==> default:  (09/14) /usr/share/moduleframework/tools/modulelint.py:DockerLint.testBasic;1:  
==> default: PASS (13.81 s)
==> default:  (10/14) /usr/share/moduleframework/tools/modulelint.py:DockerLint.testBasic;2:  
==> default: SKIP
==> default:  (11/14) /usr/share/moduleframework/tools/modulelint.py:DockerLint.testLabels;1:  
==> default: PASS (3.47 s)
==> default:  (12/14) /usr/share/moduleframework/tools/modulelint.py:DockerLint.testLabels;2:  
==> default: SKIP
==> default:  (13/14) /usr/share/moduleframework/tools/modulelint.py:ModuleLint.testPackages;1:  
==> default: PASS (13.94 s)
==> default:  (14/14) /usr/share/moduleframework/tools/modulelint.py:ModuleLint.testPackages;2:  
==> default: PASS (3.16 s)
==> default: RESULTS    : PASS 10 | ERROR 2 | FAIL 0 | SKIP 2 | WARN 0 | INTERRUPT 0
==> default: TESTS TIME : 143.11 s
==> default: JOB HTML   : /root/avocado/job-results/job-2017-02-22T13.30-a3dc381/html/results.html
```

![screenshot from 2017-02-22 14-45-35](https://cloud.githubusercontent.com/assets/8735467/23214072/bcc10bf0-f90d-11e6-9429-64a3acd92057.png)

for example output like: 
```
multiplexer values:  module:rpm
(01/02) ./generated.py:test1: PASS (34.05 s)
(01/02) ./generated.py:test2: PASS (34.05 s)

multiplexer values:  module:docker
(01/02) ./generated.py:test1: PASS (34.05 s)
(01/02) ./generated.py:test2: PASS (34.05 s)

```

**Comment by apahim** [comment]
@jscotka, in your example, you're proposing to expose the variant values as a header for the group of tests generated by that variant. This has the power to make the UI quite messy, depending on the number of parameters defined in the yaml file.

Also, the format of the test names as you see in the UI is defined by the following RFC:
https://www.redhat.com/archives/avocado-devel/2016-March/msg00024.html

There, you see that a variant has an unique ID: `A Variant ID is an arbitrary and abstract string created by the multiplexer to identify each variant.`. It means that you can easily know the values of that variant, even before running your test. For example, to see the values of the variant 12 of the `examples/mux-environment.yaml`: 

```
$ avocado multiplex -m examples/mux-environment.yaml -c
...
Variant 12:    /run/hw/cpu/amd, /run/hw/disk/scsi, /run/distro/mint, /run/env/prod
    /run/distro/mint:init       => systemv
    /run/env/prod:opt_CFLAGS    => -O2
    /run/hw/cpu/amd:cpu_CFLAGS  => -march=athlon64
    /run/hw/disk/scsi:disk_type => scsi
...
```

**Comment by jscotka** [comment]
yes exatly @apahim it is very dependent on amount of tests and values in multiplex file.
We've discussed it with @ldoktor that it could be possible to do both of them,
In case there is bit amount of variants actual solution is better, in case there is big amount of tests and just few variants in multiplex is is better to do it by my way.

**Comment by ldoktor** [comment]
Well from our discussion I understood it a bit differently and actually simpler. I thought you just want to be able to change the current tests execution from `iter-all-variants-per-test` to `iter-all-variants-on-all-tests`. The difference is:

```python
for test in tests:
    for variant in variants:
        run_test(test, variant)
```

vs.

```python
for variant in variants:
    for test in tests:
        run_test(test, variant)
```

This is actually not a new idea, we had it even on our backlog years ago but we decided to first work on the job-api which would allow this (as well as executing only some tests with variants and other without them) and then reconsider the need for this.

Anyway I'd like to be able to tell our runner to work in those two schemes as I see value in both and it's quite simple to implement (I'd even say low hanging fruit for someone who is already familiar with runners).

Note I don't think we should support all kind of combinations as that is the job-api purpose, but we should cover at least the basic variants.

Btw one of the reasons I like this other kind of iteration is that eg. `avocado-vt` tests with `keep_guest_running` would run all tests pretty quickly and only restart the guest once per all tests unlike in current situation where the HW would change between each variant. The priority for us was pretty low as we don't have such feature in `avocado-virt` and `avocado-vt` uses cartesian (which runs all tests on one variant and then all tests on other variant...)

Anyway does this makes sense? I can imagine the usage like this:

```
$ avocado run -m hw.yaml -- boot migrate
(1/4) boot;ide
(2/4) boot;ahci
(3/4) migrate;ide
(4/4) migrate;ahci

$ avocado run -m hw.yaml --mux-per-suite -- boot migrate
(1/4) boot;ide
(2/4) migrate;ide
(3/4) boot;ahci
(4/4) migrate;ahci
```

**Comment by apahim** [comment]
I see a big value in the `--mux-per-suite` approach. No question on that. My only concern here is to keep the UI test names in line with the RCF, the variant ID has to be guaranteed to be unique. But that is simple to achieve, even with `;prettyname`.

I also had the impression that @jscotka is also willing to expose the parameters of the variant in the UI. For that, no need to change anything, since we have strong logging capabilities. A test like this:
```python
import logging
from avocado import Test

LOG = logging.getLogger('params')

class MyTest(Test):

    def setUp(self):
        params = ''
        for origin, key, value in self.params.iteritems():
            params += "\n    %s:%s => %s" % (origin, key, value)
        LOG.info(params)

    def test(self):
        pass
```

Could create an output like this:
```bash
$ avocado --show app,params run ~/avocado/tests/test_expose_variant_values.py -m examples/mux-environment.yaml
JOB ID     : b411adb2ba74693a42c227ddb16adfaab8dc7d95
JOB LOG    : /home/apahim/avocado/job-results/job-2017-02-23T11.22-b411adb/job.log
 (01/24) /home/apahim/avocado/tests/test_expose_variant_values.py:MyTest.test;1:  params:
    /run/hw/cpu/intel:cpu_CFLAGS => -march=core2
    /run/hw/disk/scsi:disk_type => scsi
    /run/distro/fedora:init => systemd
    /run/env/debug:opt_CFLAGS => -O0 -g
PASS (0.02 s)
 (02/24) /home/apahim/avocado/tests/test_expose_variant_values.py:MyTest.test;2:  params:
    /run/hw/cpu/intel:cpu_CFLAGS => -march=core2
    /run/hw/disk/scsi:disk_type => scsi
    /run/distro/fedora:init => systemd
    /run/env/prod:opt_CFLAGS => -O2
PASS (0.02 s)
 (03/24) /home/apahim/avocado/tests/test_expose_variant_values.py:MyTest.test;3:  params:
    /run/hw/cpu/intel:cpu_CFLAGS => -march=core2
    /run/hw/disk/scsi:disk_type => scsi
    /run/distro/mint:init => systemv
    /run/env/debug:opt_CFLAGS => -O0 -g
PASS (0.02 s)
...
```

And with a minor change in the html plugin, we can have the `params` exposed with a mouse hover action in the test name:

![2017-02-23-112449_3840x1080_scrot](https://cloud.githubusercontent.com/assets/4717757/23255366/9882b5a0-f9bb-11e6-80d4-790a4403450e.png)

**Comment by jscotka** [comment]
Hi,
your suggestions seems well. 
I've just added there these lines of variant params, to see where one test finish and start anoher variant. It is not so imprtant to have it there, just it could improve readability in case lots of testcases and small amout of variants

**Comment by ldoktor** [comment]
OK, let me rephrase it to see whether I understand the point.

1. `--mux-per-suite` would solve your issue, right?
2. you don't really require the extra divider like `multiplexer values:  module:docker` between the variants?
3. you have no opinion regarding the `html` plugin and listing the params there?

**Comment by jscotka** [comment]
Hi,
1 yes
2 yes, not so important, just nice to have
3. yes, something similar would be nice to be able filter tests dependent on variant

4. As you mentioned, it also helps to solve situation in case I'll handle situation that I let guests/containers running between same type  tests (because faster solution).
in actual case, it is interlaced  and it is not possible to handle it

**Comment by ldoktor** [comment]
![](https://github.trello.services/images/mini-trello-icon.png) [RFC: Add a way to run full suite-per-variants instead of all-variants-of-each-test](https://trello.com/c/p51VvUnT/936-rfc-add-a-way-to-run-full-suite-per-variants-instead-of-all-variants-of-each-test)

**Comment by ldoktor** [comment]
OK, let's track this issue only here https://trello.com/c/p51VvUnT/936-rfc-add-a-way-to-run-full-suite-per-variants-instead-of-all-variants-of-each-test

**Comment by clebergnu** [comment]
@jscotka FIY, this has been merged and will be included in release 51.0.
